app/src/split/bluetooth/service.c

```diff
@@ -101,42 +99,40 @@ static ssize_t split_svc_run_behavior(struct bt_conn *conn, const struct bt_gatt
     return len;
 }
 
-static ssize_t split_svc_num_of_positions(struct bt_conn *conn, const struct bt_gatt_attr *attrs,
-                                          void *buf, uint16_t len, uint16_t offset) {
-    return bt_gatt_attr_read(conn, attrs, buf, len, offset, attrs->user_data, sizeof(uint8_t));
+static void split_svc_data_xfer_callback(struct k_work *work) {
+    LOG_DBG("Size correct, raising event");
+    struct zmk_split_data_xfer_event event;
+    memcpy(&event.data_xfer, &data_xfer_payload, sizeof(struct zmk_split_data_xfer_data));
+    raise_zmk_split_data_xfer_event(event);
```

Comment:
So... I think my main concern with this approach is that this introduces an extra level of indirection for consuming code on the peripheral sides. They no longer can subscribe to the specific event they want, they need to subscribe to this "grab bag" event and do two levels of checks/filtering.

I would much rather a system that uses some generic GATT bits as an internal implementation detail that allows domain specific events/data to be set and the rest of the code doesn't need to be aware of it. 

Does that make sense?

Reply:
So the peripheral GATT handler raises the layer changed event or indicators changed event etc? I avoided that for a couple reasons:
1. If someone wants to send their own data across this transport in a module they can't without modifying upstream files, unless we have the data transfer event as a backstop if the known tags don't match maybe
2. We'd need to create a lot more events to handle RGB indicator state, backlight state etc
3. Any alternative transports (e.g. wired) would probably need to reimplement the same sorting and individual event raising functionality, rather than just raising the same event and letting the downstream files handle it

The alternative would be having a split communication layer in between the transport's and the rest of the system on each side e.g.

System - split_central.c - (wired, Bluetooth etc etc)_central.c <-> _transport_ <-> (wired, Bluetooth etc etc)_peripherals.c - split_peripheral.c - System 

Which could pave the way for wired splits, or even having multiple different split transports available for use simultaneously, plug the trrs cable in if you're in a coffee shop or unplug it for wireless operation but that's a far more major rework than what I originally had in mind here

Reply:
Would a good path forward to be add an extra source file for a peripheral handler that captures all data transfer events and raises the appropriate events for recognised tags? Do we also want a central handler that captures events and turns that into data transfer requests. 

1: I'm very keen on maintaining support for custom tags. I'll have a think on how best to avoid ID collisions, ideally some sort of compile time check but no obvious solution for this comes immediately to mind. The results of an ID collision would also be unpredictable which would  lead to a challenging problem to diagnose. Good logging might help on this one. 

2: no, I suppose it's not. Off the top of my head we'll need new events for backlight and underglow. For the keymap it gets a little trickier, because there's the current event for layer but not the display name, which a peripheral widget may well want to display. Should we treat the keymap information as a special case and add some extra code on the peripheral side to handle storing the keymap name or add a new event on both sides for highest layer keymap's name?

Reply:
> [...] having multiple different split transports available for use simultaneously, plug the trrs cable in if you're in a coffee shop or unplug it for wireless operation but that's a far more major rework than what I originally had in mind here

Just wanted to add that this is probably the feature I'd want the most from wired split in ZMK (aside from it functioning).